### PR TITLE
[FrameworkBundle] improve AbstractController::handleForm()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.x 
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | (
| License       | MIT
| Doc PR        | -

Related to #41178

### CODE REVIEW

**Comment by lyrixx** [comment]
IMHO, this does not fix all #41178 but it helps indeed.

**Review by lyrixx — APPROVED** [review]
👍🏼 thanks (but do not close #41178 please)

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 1]
```diff
@@ -297,23 +297,30 @@ protected function stream(string $view, array $parameters = [], StreamedResponse
...
+
+            if ($submitted && 200 === $response->getStatusCode()) {
+                $response->setStatusCode(Response::HTTP_UNPROCESSABLE_ENTITY);
+            }
         }
 
-        $response = $render($form, $data);
-        if ($submitted && 200 === $response->getStatusCode()) {
-            $response->setStatusCode(Response::HTTP_UNPROCESSABLE_ENTITY);
+        if (!$response instanceof Response) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:322]: Wouldn't this break using `@Template` and other kind of view listeners? Couldn't we just check that the return value isn't null?

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 2]
```diff
@@ -297,23 +297,30 @@ protected function stream(string $view, array $parameters = [], StreamedResponse
...
-        if ($submitted && $form->isValid()) {
-            return $onSuccess($form, $data);
+
+        if ($isValid = $submitted && $form->isValid()) {
+            $response = $onSuccess($form, $data, $request);
+        } else {
+            $response = $render($form, $data, $request);
+
+            if ($submitted && 200 === $response->getStatusCode()) {
+                $response->setStatusCode(Response::HTTP_UNPROCESSABLE_ENTITY);
```
Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:318]: $response could be null, or anything

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:322]: This check should be done before checking the status code on line 317

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:322]: fixed in #41186

**Pull request merged**